engine/swarm/secrets.md

```diff
@@ -0,0 +1,770 @@
+                         -extfile "root-ca.cnf" -extensions \
+                         root_ca
+
+          Signature ok
+          subject=/C=US/ST=CA/L=San Francisco/O=Docker/CN=Swarm Secret Example CA
+          Getting Private key
+          ```
+
+    5.  Generate the site key.
+
+          ```bash
+          $ openssl genrsa -out "site.key" 4096
+
+          Generating RSA private key, 4096 bit long modulus
+          ....................................++
+          ..............++
+          e is 65537 (0x10001)
+          ```
+
+    6.  Generate the site certificate and sign it with the site key.
+
+          ```bash
+          $ openssl req -new -key "site.key" -out "site.csr" -sha256 \
+          -subj '/C=US/ST=CA/L=San Francisco/O=Docker/CN=localhost'
+          ```
+
+    7.  Configure the site certificate. Edit a new file  called `site.cnf` and
+        paste the following contents into it. This constrains the site
+        certificate so that it can only be used to authenticate a server and
+        can't be used to sign certificates.
+
+          ```none
+          [server]
+          authorityKeyIdentifier=keyid,issuer
+          basicConstraints = critical,CA:FALSE
+          extendedKeyUsage=serverAuth
+          keyUsage = critical, digitalSignature, keyEncipherment
+          subjectAltName = DNS:localhost, IP:127.0.0.1
+          subjectKeyIdentifier=hash
+          ```
+
+    8.  Sign the site certificate.
+
+          ```bash
+          $ openssl x509 -req -days 750 -in "site.csr" -sha256 \
+              -CA "root-ca.crt" -CAkey "root-ca.key"  -CAcreateserial \
+              -out "site.crt" -extfile "site.cnf" -extensions server
+
+          Signature ok
+          subject=/C=US/ST=CA/L=San Francisco/O=Docker/CN=localhost
+          Getting CA Private Key
+          ```
+
+    9.  The `csr` and `cnf` files are not needed by the Nginx service, but you
+        will need them if you want to generate a new site certificate. Protect
+        the `key` files.
+
+3.  Create three secrets, representing the key, the certificate, and the
+    `nginx.conf`. You can store any file as a secret as long as it is smaller
+    than 500 KB. This allows you to decouple the key, certificate, and
+    configuration from the services that will use them.
+
+    ```bash
+    $ docker secret create site.key -f site.key site.key
+
+    2hvoi9mnnaof7olr3z5g3g7fp
+
+    $ docker secret create -f site.crt site.crt
+
+    aya1dh363719pkiuoldpter4b
+
+    $ docker secret create -f nginx.conf nginx.conf
+
+    zoa5df26f7vpcoz42qf2csth8
+
+    $ docker secret ls
+
+    ID                          NAME                  CREATED             UPDATED
+    2hvoi9mnnaof7olr3z5g3g7fp   site.key       58 seconds ago      58 seconds ago
+    aya1dh363719pkiuoldpter4b   site.crt       24 seconds ago      24 seconds ago
+    zoa5df26f7vpcoz42qf2csth8   nginx.conf     11 seconds ago      11 seconds ago
```

Comment:
to make it easier to follow (especially with the symlink), you could consider naming the secret (e.g.) `site.conf`